Courses: Mapping a tag to something that is human-readable

sherakama opened the pull request:
# READY FOR REVIEW

@jbickar, this one is a doozy.

Steps to test:

1. Install a JSA website somewhere
2. Check out this branch
3. Enable stanford_courses_tag_translate
4. Create a course importer using http://explorecourses.stanford.edu/search?view=xml-20130201&filter-coursestatus-Active=on&page=0&catalog=&academicYear=&q=IR%3A%3Awim&collapse=
5. Find a course that was imported and review the tags that are attached
6. Go to: `admin/config/stanford/courses/tag-translate` and create a few new tag translations
7. Go back to importer and re-import
8. Go back to the course you reviewed earlier and check the human readable field for the new values.
9. Go to `admin/config/stanford/courses/tag-translate/export` and keep that information handy

### Part II: Import / Export

1. Install a new JSA website somewhere else
2. Check out this branch and enable stanford_courses_tag_translate
3. Go to `admin/config/stanford/courses/tag-translate/import` and import the values from the steps above.
4. Go to `admin/config/stanford/courses/tag-translate` and verify they imported correctly.

### Part III: Additional checks.

1. Check that the content type has the human readable field
2. Check that the feed importer has a mapping to the human readable field
3. Check that the feeds tamper has the new custom plugin on the human readable field
4. Try to import bad values.
5. Disable and uninstall to ensure that everything goes away correctly.
6. Before enabling the stanford_courses_tag_translate modify the feeds importer, feeds tampers, and content type. 

### Part IV: More things to check
1. Set the default value for the import to [original] and run an import
2. Set the default value for the import to nothing (empty field) and run an import
3. Set the default value for the import to "Some random string" and run import
4. Check to see that `admin/structure/views/view/stanford_courses_ref_human_tag` installs on enable
5. Check to see that `admin/structure/views/view/stanford_courses_ref_human_tag` uninstalls on uninstall
6. Check that Deprecated field group is on courses with two tag fields in them.

Thanks,

jbickar commented:
Yikes.

jbickar commented:
Part I passes.

jbickar commented:
Part II passes

jbickar commented:
If I have modified the Feed importer before enabling the module, then I disable and uninstall the module, the mapping and tamper remain. I'll see if I can reproduce that

jbickar commented:
OK, everything in Part III passes except when I try step 5 after step 6. Trying to reproduce just that.

jbickar commented:
Try this:

1. Add a field to the stanford_course content type (I created a long text field called "Test")
2. Map an xpath expression to that field in the Feeds importer (I used "description" for the XPath expression)
3. Create an importer node
4. Enable the stanford_courses_tag_translate module
5. Disable and uninstall the stanford_courses_tag_translate
6. Go to `admin/structure/feeds/stanford_course_importer/mapping`
7. See "Tag: Human"
8. Go to `admin/structure/feeds/stanford_course_importer/tamper`
9. See "Tag: Human" with no plugins defined

I'm verifying one more bug that I thought I found.

jbickar commented:
@sherakama see above. Passing back to you for that. I haven't reviewed the code yet; I can do that next week.

Those testing steps look like a good beginning for a Behat test :)

sherakama commented:
The feeds_tamper alters don't get written to the database when you alter the importer settings. They only get written to the db when you override them in the UI. So unless you override each and the importer they won't stick around when you disable the module. I have added some code that will remove the xpath mapping for the human field when disabled even if it is in the db. The feeds tamper items go away because it is disabled or the importer mapping does not exist.

sherakama commented:
@jbickar,

A couple more changes for you to review.
1. Changed the human field to a vocabulary
2. Added a vocabulary
3. Added some default views that get installed directly into the database on enable.
4. Moved the two tags (org/code) in to a deprecated field_group

jbickar commented:
Made a little bit of progress on testing this. The issue that I described with the XPath mapping sticking around after being disabled is resolved. Will continue testing on Monday.

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.context.inc:
```diff
@@ -0,0 +1,47 @@
+  $export = array();
+
+  $context = new stdClass();
+  $context->disabled = FALSE; /* Edit this to true to make a default context disabled initially */
```
Shouldn't this Context live in one of the Layout modules?

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.install:
```diff
@@ -0,0 +1,162 @@
+  // Check to see if the importer has been modified and is in the database.
+  $importers = ctools_export_load_object('feeds_importer', 'conditions', array('id' => 'stanford_course_importer'));
+
+  // Does not exist. This is strange but end anyway. Should we log here?
```
Hrm, good question. Without doing a lot of digging, I would assume that the `$importers["stanford_course_importer"]` object would at least contain some metadata, therefore not be empty. I think I'm fine with this just being addressed as a code comment at this point.

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.module:
```diff
@@ -0,0 +1,353 @@
+<?php
+/**
+ * @file
+ * File description
```
`@todo Describe this`

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.module:
```diff
@@ -0,0 +1,353 @@
+  switch ($path) {
+    // Main module help for the block module.
+    case 'admin/help#course-tag-translate':
+      return '<p>' . t('Ask John !email', array('!email' => l("jbickar@stanford.edu","mailto://jbickar@stanford.edu"))) . '</p>';
```
Very funny.

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.module:
```diff
@@ -0,0 +1,353 @@
+
+  // Disable the sucka so that people are not tempted to edit them.
+  $form["field_s_course_ec_tag"]["#disabled"] = TRUE;
+  $form["field_s_course_tag_human"]["#disabled"] = TRUE;
```
Oh, I see what's happening here. Hm. This UX is a little janky. We should probably hide this field on the node display, and then on the edit form, add help text indicating why they can't edit it. But if we hide the field on node display, we should probably change the term reference field from autocomplete to a different widget, so that the user can see what human-readable tags are on the course.

This work can be a separate ticket. I'll make it.
![2015-12-15-11320-screenshot](https://cloud.githubusercontent.com/assets/821106/11820960/08dc03e4-a31c-11e5-92f0-1bd049ae084e.png)
![2015-12-15-11327-screenshot](https://cloud.githubusercontent.com/assets/821106/11820961/08de2e94-a31c-11e5-9393-574818794095.png)

jbickar replied:
https://stanfordits.atlassian.net/browse/BASIC-1665

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.pages.inc:
```diff
@@ -0,0 +1,531 @@
+}
+
+/**
+ * [stanford_courses_tag_default_value_form description]
```
Yeah, this function probably should be documented.

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.pages.inc:
```diff
@@ -0,0 +1,531 @@
+  $form["ctag"] = array(
+    "#type" => 'textfield',
+    "#description" => t('The tag used on explorecourses, e.g., "REGISTRAR::training"'),
+    "#default_value" => isset($tobj->ctag) ? $tobj->ctag : "",
```
I bet @boznik put you up to this.

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.pages.inc:
```diff
@@ -0,0 +1,531 @@
+
+  // Get the submission value.
+  $values = $form_state["values"];
+  $json = filter_xss($values["txtxprt"]);
```
Of the many string sanitization functions available, why do you choose `filter_xss()` here? Asking for my own edification, not questioning your decision.

sherakama replied:
filter_xss allows for html but will strip out any nasty bits. check_plain will render html tags as plain text. We want to allow for html here.

sherakama replied:
Well, we did until this became a taxonomy field. This probably needs to go to check_plain now.

jbickar commented on modules/stanford_courses_tag_translate/stanford_courses_tag_translate.pages.inc:
```diff
@@ -0,0 +1,531 @@
+  // Stick the whole export in a textarea box and be done with it.
+  $render['theform'] = drupal_get_form("stanford_courses_tag_translate_export_form", $xport_code);
+
+  // Shares a form with the input function. Don't let them do that here.
```
LOL

jbickar commented on stanford_courses.features.field_instance.inc:
```diff
@@ -1055,7 +1055,7 @@ function stanford_courses_field_default_field_instances() {
       ),
       'type' => 'text_textfield',
-      'weight' => 17,
+      'weight' => 9,
```
We're adjusting weights of existing fields here because we're adding a field, yeah?

sherakama replied:
Yep, just more better grouping.

sherakama commented:
@jbickar,

Thanks for changing the hook_help. Were you worried that I wouldn't?

jbickar commented:
Terrified
